src/data/finsupp/convolution.lean

```diff
@@ -0,0 +1,270 @@
+
+lemma mul_def {f g : monoid_algebra k G} :
+  f * g = (f.sum $ λa₁ b₁, g.sum $ λa₂ b₂, single (a₁ * a₂) (b₁ * b₂)) :=
+rfl
+
+lemma support_mul (a b : monoid_algebra k G) :
+  (a * b).support ⊆ a.support.bind (λa₁, b.support.bind $ λa₂, {a₁ * a₂}) :=
+subset.trans support_sum $ bind_mono $ assume a₁ _,
+  subset.trans support_sum $ bind_mono $ assume a₂ _, support_single_subset
+
+/-- The unit of the multiplication is `single 1 1`, i.e. the function
+  that is `1` at `1` and zero elsewhere. -/
+instance : has_one (monoid_algebra k G) :=
+⟨single 1 1⟩
+
+lemma one_def : (1 : monoid_algebra k G) = single 1 1 :=
+rfl
+
+-- TODO: the simplifier unfolds 0 in the instance proof!
+private lemma zero_mul (f : monoid_algebra k G) : 0 * f = 0 :=
+by simp only [mul_def, sum_zero_index]
+
+private lemma mul_zero (f : monoid_algebra k G) : f * 0 = 0 :=
+by simp only [mul_def, sum_zero_index, sum_zero]
+
+private lemma left_distrib (a b c : monoid_algebra k G) : a * (b + c) = a * b + a * c :=
+by simp only [mul_def, sum_add_index, mul_add, _root_.mul_zero, single_zero, single_add,
+  eq_self_iff_true, forall_true_iff, forall_3_true_iff, sum_add]
+
+private lemma right_distrib (a b c : monoid_algebra k G) : (a + b) * c = a * c + b * c :=
+by simp only [mul_def, sum_add_index, add_mul, _root_.mul_zero, _root_.zero_mul, single_zero,
+  single_add, eq_self_iff_true, forall_true_iff, forall_3_true_iff, sum_zero, sum_add]
+
+instance : semiring (monoid_algebra k G) :=
+{ one       := 1,
+  mul       := (*),
+  one_mul   := assume f, by simp only [mul_def, one_def, sum_single_index, _root_.zero_mul,
+    single_zero, sum_zero, zero_add, one_mul, sum_single],
+  mul_one   := assume f, by simp only [mul_def, one_def, sum_single_index, _root_.mul_zero,
+    single_zero, sum_zero, add_zero, mul_one, sum_single],
+  zero_mul  := zero_mul,
+  mul_zero  := mul_zero,
+  mul_assoc := assume f g h, by simp only [mul_def, sum_sum_index, sum_zero_index, sum_add_index,
+    sum_single_index, single_zero, single_add, eq_self_iff_true, forall_true_iff, forall_3_true_iff,
+    add_mul, mul_add, add_assoc, mul_assoc, _root_.zero_mul, _root_.mul_zero, sum_zero, sum_add],
+  left_distrib  := left_distrib,
+  right_distrib := right_distrib,
+  .. finsupp.add_comm_monoid }
+
+end
+
+instance [comm_semiring k] [comm_monoid G] : comm_semiring (monoid_algebra k G) :=
+{ mul_comm := assume f g,
+  begin
+    simp only [mul_def, finsupp.sum, mul_comm],
+    rw [finset.sum_comm],
+    simp only [mul_comm]
+  end,
+  .. monoid_algebra.semiring }
+
+instance [ring k] : has_neg (monoid_algebra k G) :=
+by apply_instance
+
+instance [ring k] [monoid G] : ring (monoid_algebra k G) :=
+{ neg := has_neg.neg,
+  add_left_neg := add_left_neg,
+  .. monoid_algebra.semiring }
+
+instance [comm_ring k] [comm_monoid G] : comm_ring (monoid_algebra k G) :=
+{ mul_comm := mul_comm, .. monoid_algebra.ring}
+
+lemma single_mul_single [semiring k] [monoid G] {a₁ a₂ : G} {b₁ b₂ : k} :
```

Comment:
I'm not sure if we want to encourage working with type annotations on `finsupp.single`. We should probably have `monoid_algebra.C` and `monoid_algebra.X`. (Note that this `X` will be slightly different from the one for polynomials.)
I'm not sure if we should do this now, or in a follow-up PR.

Reply:
I'm actually not sure what you have in mind here.

Reply:
If I have a term `g : G`, how do I view this as a term of `monoid_algebra k G`? Currently you have to use `finsupp.single` to do this. I think we should have a dedicated definition for this, to guide type inference in the right direction, and for better readability. (Even in maths we usually denote these elements with `X_g` or `X^g` or `[g]` or something like that.)

Reply:
It's the same reason why we define `mv_polynomial.monomial` instead of using `finsupp.single` there.

Reply:
Sure, I get it. Shall we just introduce `monoid_algebra.single`? There is a cost to this sort of thing, however: since `simp` works syntactically you may have to restate lots of lemmas. I wonder if making it an `abbreviation` gives you the advantages of guiding type inference still.

Reply:
If `abbreviation` gives us the best of both worlds, I would certainly vote for that. I'm not sure if I like `single`. It seems to do two things at once: both the constants (giving the algebra structure) and the generators. I would rather separate them. But like I said before, maybe that shouldn't be done in this PR.

Reply:
Okay. My preference is not to introduce a `monoid_algebra.single` in this PR; I'm focussing on other things at the moment.